src/module.cpp

```diff
@@ -41,12 +41,83 @@
 #include <set>
 #include <utility>
 #include <unordered_set>
+#include <bitset>
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
 MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_TRACE_FINALIZE)
 
+template <std::size_t N>
+struct bit_signal
+{
+    std::bitset<N> slots;
+    std::bitset<N> allocated;
```

Comment:
why do you need bitset<N>. 

Why can't simple `bool` flag is is enough ? 
Is it because of multi-threaded parallel compilation? Then is `64` sufficient ?

Reply:
There could be multiple subscribers withing the same thread, especially since this is called recursively. Once for the optimize_module loop and then again for the sub-passes like simplify_algebra, simplify_reshapes, etc.

> Is it because of multi-threaded parallel compilation?

The current implementation is not really thread-safe, and it seems unlikely to be used across threads since modifying the module across threads is not thread-safe either.